Code to write data in respective Column in case of XDDFChart

sandeeptiwari32 opened the pull request:
updated code of XDDFChart method fillsheet, added a variable to identify the column index and write data in respective index in embedded excel file

sandeeptiwari32 committed:
updated code for fill sheet

sandeeptiwari32 committed:
Example file editing chart data for docx

created example class file to modify chart data

sandeeptiwari32 committed:
update code for fill sheet method data

update code for fill sheet method data

sandeeptiwari32 committed:
updated fill sheet method

updated fill sheet method

sandeeptiwari32 committed:
updated space and tab

updated space and tab

sandeeptiwari32 committed:
update space and tab

update space and tab

sandeeptiwari32 commented:
this commit also contain one example file same as we have for XSLFChart

sandeeptiwari32 reviewed (commented):
updated code of XDDFChart method fillsheet, added a variable to identify the column index and write data in respective index in embedded excel file

asfgit commented:
Can one of the admins verify this patch?

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -75,8 +76,12 @@
      */
     private XSSFWorkbook workbook;
 
+    XSSFSheet sheet = null;
```
`private`

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -75,8 +76,12 @@
+    
     private int chartIndex = 0;
 
+    POIXMLDocumentPart documentPart = null;
```
`private`

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -460,12 +465,46 @@ public void saveWorkbook(XSSFWorkbook workbook) throws IOException, InvalidForma
-            row.createCell(1).setCellValue(valuesData.getPointAt(i).doubleValue());
+            XSSFRow row = this.getRow(sheet,i + 1); // first row is for title
+            this.getCell(row,categoryData.getColIndex()).setCellValue(categoryData.getPointAt(i).toString());
+            this.getCell(row,valuesData.getColIndex()).setCellValue(valuesData.getPointAt(i).doubleValue());
```
Space after `,` helps for code readability.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -460,12 +465,46 @@ public void saveWorkbook(XSSFWorkbook workbook) throws IOException, InvalidForma
+     * @since POI 4.0.0
+     */
+    private XSSFRow getRow(XSSFSheet sheet,int index)
+    {
```
Open `{` at end of previous line.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -460,12 +465,46 @@ public void saveWorkbook(XSSFWorkbook workbook) throws IOException, InvalidForma
+     */
+    private XSSFRow getRow(XSSFSheet sheet,int index)
+    {
+        if(sheet.getRow(index)!=null)
```
`if (sheet.getRow(index) != null)` is more readable with the spaces.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -527,26 +566,14 @@ public String formatRange(CellRangeAddress range) {
+        if(sheet==null)
+        {
+            workbook = new XSSFWorkbook();
+            sheet = workbook.createSheet();
```
This seems to remove the ability to modify an existing chart.

sandeeptiwari32 replied:
we are copying data from old sheet to this new sheet (in case of modifying chart data) in fillSheet() method, so i think this will not break any existing functionality.
we need to create new sheet because when we are writing embedded xlsx part outputstrem which we read,when writing it  in document it always writing old data into embedded excel.

sandeeptiwari32 commented:
@Alain-Bearez sure I will do....

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 reviewed (commented):
updated suggested changes.

Alain-Bearez commented on src/examples/src/org/apache/poi/xwpf/usermodel/examples/BarChartExampleDOCX.java:
```diff
@@ -0,0 +1,143 @@
+ * ====================================================================
+ */
+
+package org.apache.poi.xwpf.usermodel;
```
This does not compile since the package does not match the folder structure.

Alain-Bearez commented on src/examples/src/org/apache/poi/xwpf/usermodel/examples/BarChartExampleDOCX.java:
```diff
@@ -0,0 +1,143 @@
+        final String valuesDataRange2 = chart.formatRange(new CellRangeAddress(1, numOfPoints, 2, 2));
+        final XDDFDataSource<?> categoriesData = XDDFDataSourcesFactory.fromArray(categories, categoryDataRange,0);
+        final XDDFNumericalDataSource<? extends Number> valuesData = XDDFDataSourcesFactory.fromArray(values, valuesDataRange,1);
+        values[3]=(double) 10;
```
```
        values[2] = 10.0;
```

I stumbled upon this code smell since it was failing at execution.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -460,12 +465,44 @@ public void saveWorkbook(XSSFWorkbook workbook) throws IOException, InvalidForma
-            XSSFRow row = sheet.createRow(i + 1); // first row is for title
-            row.createCell(0).setCellValue(categoryData.getPointAt(i).toString());
-            row.createCell(1).setCellValue(valuesData.getPointAt(i).doubleValue());
+            XSSFRow row = this.getRow(sheet,i + 1); // first row is for title
```
I will open a pull request to your branch to fix these formatting details. It will be easier than trying to comment all of them.

Alain-Bearez commented on src/examples/src/org/apache/poi/xwpf/usermodel/examples/BarChartExampleDOCX.java:
```diff
@@ -0,0 +1,143 @@
+            try (XWPFDocument doc = new XWPFDocument(argIS)) {
+                XWPFChart chart = findChart(doc);
+                setBarData(chart, chartTitle, categories, values);
+                setColumnData(chart, "Column variant");
```
In the original example, two charts were created from a template with only one of them. One of them was populated by `setBarData` while the other was populated by `setColumnData`.

In your example, you start from a template with two charts and only change the first one, applying both methods...

May I suggest that you would experiment more of your new API changes with the former approach?

Alain-Bearez committed:
fix formatting and details from previous reviews

Alain-Bearez reviewed (changes requested):
I will push a pull request to your branch within some minutes to resolve my comments.

sandeeptiwari32 committed:
Merge pull request #2 from apache/trunk

update checkout

sandeeptiwari32 committed:
Merge pull request #1 from cuali/column-index

fix formatting and details from previous reviews

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 committed:
Updated suggested changes

Updated suggested changes

sandeeptiwari32 committed:
updated code to set workbook.

in final writing of chart we are writing workbook part so instead of calling saveWorkbook method we are now calling setWorkbook method to initialise workbook object.

sandeeptiwari32 committed:
updated getsheet method

updated getsheet method

sandeeptiwari32 reviewed (commented):
updated suggested code and formatting.Also reverted code of creating new workbook when we have embedded workbook object.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -75,9 +83,11 @@
      */
     private XSSFWorkbook workbook;
-
+    
```
once more: check your configuration to not have spaces on empty lines

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -537,10 +581,35 @@ protected void commit() throws IOException {
+    private void updateSheetTable(CTTable ctTable, String title, int index) {
+        CTTableColumns tableColumnList = ctTable.getTableColumns();
+        CTTableColumn column = null;
+        if(tableColumnList.getCount() >= index)
```
Sorry to insist about code formatting... but 
```
        if (tableColumnList.getCount() >= index) {
            column = tableColumnList.getTableColumnArray(index);
        } else {
            column =  tableColumnList.addNewTableColumn();
            column.setId(index);
        }
```

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -631,7 +695,19 @@ public XSSFWorkbook getWorkbook() throws IOException, InvalidFormatException {
+            fos = new FileOutputStream(file);
+            workbook.write(fos);
+            fos.close();
+            this.workbook = new XSSFWorkbook(file);
```
Why writing the given workbook to a file?
Why using a file which is outside of the OPC package?
Why creating a new workbook from the same content?

sandeeptiwari32 replied:
i tried to use existing workbook and we are able to successfully update workbook sheet data but at the time of writing embedded part it is writing old input-stream data and due to that when we are modifying any property from word/ppt it showing old data which we read at the time of reading  word/ppt file. so what i found the input-stream writing old data and due to that i first write old data into temp-file and then read that file.

Alain-Bearez replied:
This code was trying to work around some bug in XSSFWorkbook. Now the workaround has been officially committed to the trunk. According to recent conversation on the POI Developers List, the origin of the issue with not correctly writing only the new values in the workbook has been fixed, so that this work around would not be required anymore.

pjfanning replied:
I undid this on trunk just now (tests still pass)

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFDataSourcesFactory.java:
```diff
@@ -194,6 +228,10 @@ public void setFormatCode(String formatCode) {
         }
+
+        public StringArrayDataSource(String[] elements, String dataRange, int col) {
+            super(elements, dataRange,col);
```
That one missing space after comma did escape my previous review, sorry!

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 reviewed (commented):
updated suggested changes
